### PR TITLE
Fix graphql concurrency issue with ContentPickerFiled

### LINKED ISSUE #15731: [GraphQL] A concurrent update was performed on this collection and corrupted its state.
When using GraphiQL editor, sometimes I'm getting this exception, when querying content items with nested lists.
It's not consistently repeatable for me (approx once every five queries).

image

Stacktrace:

```
GraphQL.Execution.UnhandledError: Error trying to resolve field 'contentItems'.
 ---> System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at YesSql.Session.Get[T](IList`1 documents, String collection)
   at YesSql.Services.DefaultQuery.Query`1.ListImpl()
   at YesSql.Services.DefaultQuery.Query`1.ListImpl()
   at OrchardCore.ContentManagement.GraphQL.DataLoaderExtensions.LoadPublishedContentItemsAsync(IEnumerable`1 contentItemIds, ISession session)
   at GraphQL.DataLoader.CollectionBatchDataLoader`2.FetchAsync(IEnumerable`1 list, CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderPair`2.GetResultAsync(CancellationToken cancellationToken) in /_/src/GraphQL.DataLoader/DataLoaderPair.cs:line 81
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderPair`2.GetResultAsync(CancellationToken cancellationToken) in /_/src/GraphQL.DataLoader/DataLoaderPair.cs:line 81
   at GraphQL.DataLoader.DataLoaderPair`2.GetResultAsync(CancellationToken cancellationToken) in /_/src/GraphQL.DataLoader/DataLoaderPair.cs:line 81
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderExtensions.<>c__DisplayClass5_0`2.<<Then>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   at GraphQL.DataLoader.SimpleDataLoader`1.GraphQL.DataLoader.IDataLoaderResult.GetResultAsync(CancellationToken cancellationToken)
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 515
   --- End of inner exception stack trace ---
```

@sebastienros I've just found that you fixed [similar ](https://github.com/OrchardCMS/OrchardCore/issues/9651) issue 3 years ago.

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #15731

### CODE REVIEW

**Comment by sebastienros** [comment]
@lampersky Can you check this PR?

**Comment by lampersky** [comment]
@sebastienros I've just checked, and I'm still getting exception from time to time:

```
GraphQL.Execution.UnhandledError: Error trying to resolve field 'contentItems'.
 ---> System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at YesSql.Session.Get[T](IList`1 documents, String collection)
   at YesSql.Services.DefaultQuery.Query`1.ListImpl()
   at YesSql.Services.DefaultQuery.Query`1.ListImpl()
   at OrchardCore.ContentManagement.GraphQL.DataLoaderExtensions.LoadPublishedContentItemsAsync(IEnumerable`1 contentItemIds, ISession session)
   at GraphQL.DataLoader.CollectionBatchDataLoader`2.FetchAsync(IEnumerable`1 list, CancellationToken cancellationToken)
   at GraphQL.DataLoader.DataLoaderPair`2.GetResultAsync(CancellationToken cancellationToken) in /_/src/GraphQL.DataLoader/DataLoaderPair.cs:line 81
   at OrchardCore.ContentFields.GraphQL.SebContentPickerFieldQueryObjectType.<>c.<<-ctor>b__0_1>d.MoveNext() in C:\\repo\\Project.Core\\SebFix.cs:line 34
--- End of stack trace from previous location ---
   at GraphQL.Resolvers.FuncFieldResolver`2.<>c__DisplayClass3_0.<<-ctor>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at GraphQL.Execution.ExecutionStrategy.ExecuteNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 472
   --- End of inner exception stack trace ---
```

image

**Comment by Piedone** [comment]
@coderabbit review

**Comment by Piedone** [comment]
Ooops, wrong CodeRabbit. @coderabbitai review

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/ContentPickerFieldQueryObjectType.cs [hunk 1]
```diff
@@ -27,14 +27,13 @@ public ContentPickerFieldQueryObjectType()
...
-                .ResolveAsync(x =>
+                .ResolveLockedAsync(async x =>
                 {
                     var contentItemLoader = x.GetOrAddPublishedContentItemByIdDataLoader();
 
-                    return (contentItemLoader.LoadAsync(x.Page(x.Source.ContentItemIds))).Then(itemResultSet =>
-                    {
-                        return itemResultSet.SelectMany(x => x);
-                    });
+                    var data = await contentItemLoader.LoadAsync(x.Page(x.Source.ContentItemIds)).GetResultAsync();
```
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/ContentPickerFieldQueryObjectType.cs:34]: You think about `.Result` probably. This is fine.

**Pull request closed without merging**